--- moria-5.5.2.orig/source/misc3.c	Fri Jul 22 03:47:30 1994
+++ moria-5.5.2/source/misc3.c	Fri Oct 12 21:00:00 2001
@@ -1249,8 +1249,7 @@
 {
   int i;
   register inven_type *i_ptr;
-  vtype prt2;
-  bigvtype prt1;
+  bigvtype prt1, prt2;
 #ifdef ATARIST_MWC
   int32u holder;
 #endif
--- moria-5.5.2.orig/source/misc4.c	Fri Jul 22 03:47:31 1994
+++ moria-5.5.2/source/misc4.c	Fri Oct 12 21:00:00 2001
@@ -33,7 +33,7 @@
 void scribe_object()
 {
   int item_val, j;
-  vtype out_val, tmp_str;
+  bigvtype out_val, tmp_str;
 
   if (inven_ctr > 0 || equip_ctr > 0)
     {
--- moria-5.5.2.orig/source/moria1.c	Fri Jul 22 03:47:33 1994
+++ moria-5.5.2/source/moria1.c	Fri Oct 12 21:00:00 2001
@@ -332,8 +332,8 @@
 	{
 	  objdes(tmp_val, &inventory[i], TRUE);
 	  tmp_val[lim] = 0;	 /* Truncate if too long. */
-	  (void) sprintf(out_val[i], "  %c) %s", 'a'+i, tmp_val);
-	  l = strlen(out_val[i]);
+	  (void) sprintf(out_val[i], "%c) %s", 'a'+i, tmp_val);
+	  l = strlen(out_val[i]) + 2;
 	  if (weight)
 	    l += 9;
 	  if (l > len)
@@ -352,9 +352,12 @@
 	{
 	  /* don't need first two spaces if in first column */
 	  if (col == 0)
-	    prt(&out_val[i][2], current_line, col);
-	  else
 	    prt(out_val[i], current_line, col);
+	  else
+	    {
+	      put_buffer("  ", current_line, col);
+	      prt(out_val[i], current_line, col+2);
+	    }
 	  if (weight)
 	    {
 	      total_weight = inventory[i].weight*inventory[i].number;
@@ -466,9 +469,9 @@
 	    }
 	  objdes(prt2, &inventory[i], TRUE);
 	  prt2[lim] = 0; /* Truncate if necessary */
-	  (void) sprintf(out_val[line], "  %c) %-14s: %s", line+'a',
+	  (void) sprintf(out_val[line], "%c) %-14s: %s", line+'a',
 			 prt1, prt2);
-	  l = strlen(out_val[line]);
+	  l = strlen(out_val[line]) + 2;
 	  if (weight)
 	    l += 9;
 	  if (l > len)
@@ -488,9 +491,12 @@
 	{
 	  /* don't need first two spaces when using whole screen */
 	  if (col == 0)
-	    prt(&out_val[line][2], line+1, col);
-	  else
 	    prt(out_val[line], line+1, col);
+	  else
+	    {
+	      put_buffer("  ", line+1, col);
+	      prt(out_val[line], line+1, col+2);
+	    }
 	  if (weight)
 	    {
 	      total_weight = i_ptr->weight*i_ptr->number;
